src/test_utils.rs

```diff
@@ -0,0 +1,135 @@
+//! A set of test utilities.
+//!
+//! There is some overlap between the `test_utils::write_...` functions and the code in the
+//! `encoder.rs` module.  The main difference is that the `test_utils` function perform no
+//! validation - this allows testcases to construct arbitrary, potentially invalid PNG inputs.
+//! OTOH, we should try to reuse code whenever possible - e.g.  `test_utils::write_chunk` is just a
+//! thin wrapper around `encoder::write_chunk`.
+
+use byteorder::WriteBytesExt;
+use std::convert::TryInto;
+use std::io::Write;
+
+/// Generates a store-only, non-compressed image:
+///
+/// * `00` compression mode (i.e.`BTYPE` = `00` = no compression) is used
+/// * No filter is applied to the image rows
+///
+/// Currently the image always has the following properties:
+///
+/// * Single `IDAT` chunk
+/// * Zlib chunks of maximum possible size
+/// * 8-bit RGBA
+///
+/// These images are somewhat artificial, but may be useful for benchmarking performance of parts
+/// outside of `fdeflate` crate and/or the `unfilter` function (e.g. these images were originally
+/// used to evaluate changes to minimize copying of image pixels between various buffers - see
+/// [this
+/// discussion](https://github.com/image-rs/image-png/discussions/416#discussioncomment-7436871)
+/// for more details).
+#[allow(dead_code)] // Used from `benches/decoder.rs`
+pub fn write_noncompressed_png(w: &mut impl Write, width: u32) {
+    write_png_sig(w);
+    write_rgba8_ihdr_with_width(w, width);
+    write_rgba8_idat_with_width(w, width);
+    write_iend(w);
+}
+
+/// Writes PNG signature.
+/// See http://www.libpng.org/pub/png/spec/1.2/PNG-Structure.html#PNG-file-signature
+pub fn write_png_sig(w: &mut impl Write) {
+    const SIG: [u8; 8] = [137, 80, 78, 71, 13, 10, 26, 10];
+    w.write_all(&SIG).unwrap();
+}
+
+/// Writes an arbitrary PNG chunk.
+pub fn write_chunk(w: &mut impl Write, chunk_type: &[u8], data: &[u8]) {
+    let chunk_type = crate::chunk::ChunkType(
+        chunk_type
+            .try_into()
+            .expect("`chunk_type` needs to be 4 bytes long"),
+    );
+    crate::encoder::write_chunk(w, chunk_type, data).unwrap();
```

Comment:
Reusing `write_chunk` is probably more trouble than it is worth, since this file is both a submodule of the crate when `#[cfg(test)]` is enabled and is external to the crate when used from the benchmarks

Reply:
Ooops... I didn't notice that my latest commit broke building of `bench/decoder.rs`.  Fixed.